# FileIO & Exceptions
a file is a contiguous set of bytes used to store data. This data is organized in a specific format and can be anything as simple as a text file or as complicated as a program executable. In the end, these byte files are then translated into binary 1 and 0 for easier processing by the computer.

Files on most modern file systems are composed of three main parts:

+ Header: metadata about the contents of the file (file name, size, type, and so on)
+ Data: contents of the file as written by the creator or editor
+ End of file (EOF): special character that indicates the end of the file
## Opening and Closing a File in Python
When you want to work with a file, the first thing to do is to open it. This is done by invoking the open() built-in function. open() has a single required argument that is the path to the file. open() has a single return, the file object:

file = open('dog_breeds.txt') 

When you’re manipulating a file, there are two ways that you can use to ensure that a file is closed properly, even when encountering an error. The first way to close a file is to use the try-finally block:

reader = open('dog_breeds.txt')
try:
     
finally:
    
    reader.close()

![](https://pynative.com/wp-content/uploads/2021/06/python-file-open.png)  

Other options for modes  most commonly used ones are the following:



![](https://images.slideplayer.com/64/11811821/slides/slide_6.jpg)

### Text File Types
A text file is the most common file that you’ll encounter. Here are some examples of how these files are opened:

open('abc.txt')

open('abc.txt', 'r')

open('abc.txt', 'w')
With these types of files, open() will return a TextIOWrapper file object:

>>> file = open('dog_breeds.txt')
>>> type(file)
<class '_io.TextIOWrapper'>
This is the default file object returned by open().

### Buffered Binary File Types
A buffered binary file type is used for reading and writing binary files. Here are some examples of how these files are opened:

open('abc.txt', 'rb')

open('abc.txt', 'wb')
With these types of files, open() will return either a BufferedReader or BufferedWriter file object:

>>> file = open('dog_breeds.txt', 'rb')
>>> type(file)
<class '_io.BufferedReader'>
>>> file = open('dog_breeds.txt', 'wb')
>>> type(file)
<class '_io.BufferedWriter'>

### Raw File Types
A raw file type is:

“generally used as a low-level building-block for binary and text streams.”

It is therefore not typically used.

Here’s an example of how these files are opened:

open('abc.txt', 'rb', buffering=0)
With these types of files, open() will return a FileIO file object:

>>> file = open('dog_breeds.txt', 'rb', buffering=0)
>>> type(file)
<class '_io.FileIO'>

### Tips and Tricks
Now that you’ve mastered the basics of reading and writing files, here are some tips and tricks to help you grow your skills.

__file__
The __file__ attribute is a special attribute of modules, similar to __name__. It is:

“the pathname of the file from which the module was loaded, if it was loaded from a file.”

### Appending to a File
Sometimes, you may want to append to a file or start writing at the end of an already populated file. This is easily done by using the 'a' character for the mode argument:

with open('dog_breeds.txt', 'a') as a_writer:
    a_writer.write('\nBeagle')

    

## Exceptions
Exceptions versus Syntax Errors
  Syntax errors occur when the parser detects an incorrect statement. Observe the following example:

>>> print( 0 / 0 ))
  File "<stdin>", line 1
    print( 0 / 0 ))
                  ^
SyntaxError: invalid syntax 





#### Raising an Exception
We can use raise to throw an exception if a condition occurs. The statement can be complemented with a custom exception.
![](https://files.realpython.com/media/raise.3931e8819e08.png)

Illustration of  raise statement usage
If you want to throw an error when a certain condition occurs using raise, you could go about it like this:

 >>> x = 10
 >>> if x > 5
  >>> raise Exception('x should not exceed 5. The value of x was: {}'.format(x)) 


#### The AssertionError Exception
Instead of waiting for a program to crash midway, you can also start by making an assertion in Python. We assert that a certain condition is met. If this condition turns out to be True, then that is excellent! The program can continue. If the condition turns out to be False, you can have the program throw an AssertionError exception.
![](https://files.realpython.com/media/assert.f6d344f0c0b4.png)

Python assert statement
Have a look at the following example, where it is asserted that the code will be executed on a Linux system:

>>>import sys
>>> assert ('linux' in sys.platform), "This code runs on Linux only."


#### The try and except Block: Handling Exceptions
The try and except block in Python is used to catch and handle exceptions. Python executes code following the try statement as a “normal” part of the program. The code that follows the except statement is the program’s response to any exceptions in the preceding try clause.
![](https://files.realpython.com/media/try_except.c94eabed2c59.png)

Diagram showing try and except statements
As you saw earlier, when syntactically correct code runs into an error, Python will throw an exception error. This exception error will crash the program if it is unhandled. The except clause determines how your program responds to exceptions.

The following function can help you understand the try and except block:

def linux_interaction():
    assert ('linux' in sys.platform), "Function can only run on Linux systems."
    print('Doing something.')


#### The else Clause
In Python, using the else statement, you can instruct a program to execute a certain block of code only in the absence of exceptions.

![](https://files.realpython.com/media/try_except_else.703aaeeb63d3.png)

Diagram of try, except, and else statements in Python
Look at the following example:

>>> try:
  
  >>>linux_interaction()

>>>except AssertionError as error:

>>>    print(error)

>>>else:

>>>    print('Executing the else clause.')

If you were to run this code on a Linux system, the output would be the following:

Doing something.
Executing the else clause.
Because the program did not run into any exceptions, the else clause was executed.


### Cleaning Up After Using finally
Imagine that you always had to implement some sort of action to clean up after executing your code. Python enables you to do so using the finally clause.
![](https://files.realpython.com/media/try_except_else_finally.a7fac6c36c55.png)

Diagram explaining try except else finally statements
Have a look at the following example:

>>>try:

   >>>linux_interaction()

>>>except AssertionError as error:

 >>> print(error)

>>>else:

   >>>try:

   >>>with open('file.log') as file:

  >>>read_data = file.read()

>>>except FileNotFoundError as fnf_error:

 >>>print(fnf_error)

>>>finally:

 >>>print('Cleaning up, irrespective of any exceptions.')


In the previous code, everything in the finally clause will be executed. It does not matter if you encounter an exception somewhere in the try or else clauses. Running the previous code on a Windows machine would output the following:

Function can only run on Linux systems.
Cleaning up, irrespective of any exceptions.

## Excptions summary :
+ raise allows you to throw an exception at any time.
+ assert enables you to verify if a certain condition is met and throw an exception if it isn’t.
+ In the try clause, all statements are executed until an exception is encountered.
+ except is used to catch and handle the exception(s) that are encountered in the try clause.
+ else lets you code sections that should run only when no exceptions are encountered in the try clause.
+ finally enables you to execute sections of code that should always run, with or without any previously encountered exceptions.